File: solutions/greedy/402-remove-k-digits.py
I'll help you create a solution for the "Remove K Digits" problem. This implementation will create the smallest possible number by removing k digits from a given number string.

```python
#!/usr/bin/env python3
"""
Remove K Digits - Python Implementation

This module provides a solution to remove k digits from a number string
to create the smallest possible number.

Example:
    num = "1432219", k = 3
    Result: "1219" (after removing 4, 3, 2)
"""


def remove_k_digits(num: str, k: int) -> str:
    """
    Remove k digits from the input number to get the smallest possible number.
    
    Args:
        num: Input number as string
        k: Number of digits to remove
    
    Returns:
        String representing the smallest possible number after removing k digits
    """
    # Handle edge cases
    if k >= len(num):
        return "0"
    
    if k == 0:
        return num
    
    # Use stack to keep track of digits
    stack = []
    
    # Process each digit
    for digit in num:
        # Remove digits from stack while we can and current digit is smaller
        while k > 0 and stack and stack[-1] > digit:
            stack.pop()
            k -= 1
        stack.append(digit)
    
    # If we still need to remove digits, remove from the end
    while k > 0:
        stack.pop()
        k -= 1
    
    # Construct the result
    result = ''.join(stack).lstrip('0')
    
    # Return "0" if result is empty, otherwise return result
    return result if result else "0"


def main():
    """
    Main function to demonstrate the usage of remove_k_digits function.
    """
    # Test cases
    test_cases = [
        ("1432219", 3),
        ("10200", 1),
        ("10", 2),
        ("112", 1),
        ("1234567890", 9)
    ]
    
    # Run test cases
    for num, k in test_cases:
        result = remove_k_digits(num, k)
        print(f"Input: num = {num}, k = {k}")
        print(f"Output: {result}")
        print("-" * 30)


if __name__ == "__main__":
    main()
```

This implementation includes:

1. A main function `remove_k_digits` that implements the core logic using a stack-based approach
2. Proper handling of edge cases
3. Clear documentation and comments
4. A demonstration main function with test cases
5. Type hints for better code readability
6. Proper Python conventions and best practices

The algorithm works by:
1. Using a stack to maintain digits
2. Processing each digit and removing larger digits when possible
3. Handling leading zeros
4. Managing edge cases appropriately

The time complexity is O(n) where n is the length of the input number string, and the space complexity is O(n) for the stack.

You can run this file directly to see the test cases in action. The implementation is efficient and handles various edge cases properly.

File: solutions/greedy/402-remove-k-digits.js
I'll help you create a solution for the "Remove K Digits" problem. I'll write a complete JavaScript implementation that follows best practices and includes detailed comments.

```javascript
/**
 * Remove K Digits
 * Given a non-negative integer num represented as a string and an integer k,
 * return the smallest possible integer after removing k digits from num.
 * 
 * @param {string} num - The input number as a string
 * @param {number} k - Number of digits to remove
 * @return {string} - The smallest possible number after removing k digits
 */
function removeKdigits(num, k) {
    // Handle edge cases
    if (k >= num.length) return "0";
    if (k === 0) return num;

    // Use stack to build result
    const stack = [];
    
    // Process each digit
    for (let digit of num) {
        // While we can still remove digits and current digit is smaller than last digit in stack
        while (k > 0 && stack.length > 0 && stack[stack.length - 1] > digit) {
            stack.pop();
            k--;
        }
        stack.push(digit);
    }
    
    // If we still need to remove digits, remove from end
    while (k > 0) {
        stack.pop();
        k--;
    }
    
    // Build result string
    let result = stack.join('');
    
    // Remove leading zeros
    result = result.replace(/^0+/, '');
    
    // Return "0" if empty string, otherwise return result
    return result === '' ? '0' : result;
}

/**
 * Test cases to verify the implementation
 */
function runTests() {
    const testCases = [
        { num: "1432219", k: 3, expected: "1219" },
        { num: "10200", k: 1, expected: "200" },
        { num: "10", k: 2, expected: "0" },
        { num: "112", k: 1, expected: "11" },
        { num: "1234567", k: 3, expected: "1234" },
        { num: "9", k: 1, expected: "0" },
        { num: "10001", k: 4, expected: "0" }
    ];

    testCases.forEach((test, index) => {
        const result = removeKdigits(test.num, test.k);
        console.log(`Test ${index + 1}:`);
        console.log(`Input: num = "${test.num}", k = ${test.k}`);
        console.log(`Expected: "${test.expected}"`);
        console.log(`Actual: "${result}"`);
        console.log(`Result: ${result === test.expected ? 'PASS' : 'FAIL'}\n`);
    });
}

// Export the function for potential module usage
module.exports = {
    removeKdigits
};

// Run tests if file is executed directly
if (require.main === module) {
    runTests();
}
```

This implementation includes:

1. A main `removeKdigits` function that solves the problem using a stack-based approach
2. Comprehensive comments explaining the logic and implementation
3. Edge case handling
4. A test suite with various test cases
5. Proper module exports for reusability
6. Clean code structure following JavaScript conventions

The algorithm works by:
1. Using a stack to build the result
2. Processing each digit and removing larger digits when possible
3. Handling remaining removals if needed
4. Cleaning up leading zeros
5. Returning the final result

The time complexity is O(n) where n is the length of the input string, and space complexity is O(n) for the stack.

The code includes test cases to verify the implementation works correctly for various scenarios, including edge cases.

To use this code, you can either:
1. Import it as a module in another file
2. Run it directly to see the test results
3. Call the `removeKdigits` function with your own inputs